### ISSUE TITLE
Possible deadlock in flow-ingester when creating multiple partitions in a single transaction

### ISSUE THREAD

**Issue by psFried** [estuary/flow#92]
I think I managed to wedge the ingester.

I used curl to send a request to `/ingest` that includes about 8 documents, 4 of which will require new partitions to be created due to having  new unique values for the field we are partitioned on. I did see one journal spec appear after doing a `gazctl journals list`, so I ran `gazctl journals apply` to edit that spec, which succeeded. Looking back at the curl request to `/ingest`, it had now returned the error: 
```
< HTTP/1.1 400 Bad Request
< Content-Type: text/plain; charset=utf-8
< X-Content-Type-Options: nosniff
< Date: Thu, 04 Feb 2021 20:17:44 GMT
< Content-Length: 83
<
creating journal 'segment/raw-events/type=alias/pivot=00': ETCD_TRANSACTION_FAILED
```

I re-tried the curl request to `/ingest`, but none of the other new partitions appeared in subsequent listings of the journal specs. This time the request hangs (expected, due to the storage root directory not being set), but the new journal specs are never created. I then captured the goroutine traces here:

[goroutines.txt](https://github.com/estuary/flow/files/5928542/goroutines.txt)

I'm seeing that here: 
```
1 @ 0x5c2ea0 0x591735 0x5913eb 0xd339bb 0xd3402b 0xd56498 0xd5623d 0xd59fb1 0x8bc774 0xd3a772 0x8be5b5 0x8bfb03 0x8bb56c 0x5f2f71
#	0xd339ba	github.com/estuary/flow/go/flow.(*Ingestion).Prepare+0x1ea		/home/runner/work/flow/flow/go/flow/ingest.go:198
```

it's waiting on `var pub = <-i.ingester.pubCh` (ingest.go:198).

But over at:
```
1 @ 0x5c2ea0 0x5d3001 0xd3760e 0xce46a3 0xce4109 0x5f2f71
#	0xd3760d	github.com/estuary/flow/go/flow.(*Ingester).QueueTasks.func1+0x34d	/home/runner/work/flow/flow/go/flow/ingest.go:102
``` 

It's waiting on:
```
select {
case pub = <-i.pubCh: // Pass.
case <-tasks.Context().Done():
	return nil
}
```

[ingesters.log](https://github.com/estuary/flow/files/5928533/ingesters.log)

At this point, I get the same result every time on subsequent requests to `/ingest` that add events. These requests are all going to the same pod. If I delete that pod and port-forward to another one, then I'm able to follow all these steps again and get the same result. One partition is created, then I see `ETCD_TRANSACTION_FAILED`, and then subsequent requests get deadlocked with the same goroutine traces.

**Comment by jgraettinger** [comment]
Code spelunking, the hang is expected behavior (at least right now), because once any Ingestion.Prepare() returns an error that occurs after it obtains the single `ingestionPublisher`, it doesn't return that instance and all future Prepares will block forever.

The `ingestionPublisher` _is_ in an unrecoverable state -- it's written some but not all uncommitted messages; we can't complete the current transaction but `ingestionPublisher` isn't intended to offer roll-back itself (rather, you crash and your successor process starts with a roll-back).

These aren't (supposed to be) errors which can happen due to user code, but only due to configuration or cluster environment issues which require operator intervention. If these were consumer shards (and we're moving towards this model with ingestion) then we'd fail the shard and report the error through e.x. `gazctl`.

So the issue here is really just that the error isn't bubbling up to a process crash. I'm looking at an update to mark the `ingestionPublisher` as failed but return it to its channel, and from there to bubble up to a proper crash.

The remaining question is: what caused the original `creating journal 'segment/raw-events/type=alias/pivot=00': ETCD_TRANSACTION_FAILED`. That's coming from [this guard](https://github.com/estuary/flow/blob/master/go/flow/mapping.go#L80-L87), which is supposed to bail out if the Mapping has now made **more than one attempt** to create the journal and received a "transaction failed" message. But looking at it now, there's a bug here -- we're not actually blocking until we've caught up with the future, conflicting revision !

To confirm the failure mode, can you say more about how these requests were made ? Were they separate requests, and were there more than one flow-ingester processes running who might have received them?

**Comment by jgraettinger** [comment]
| To confirm the failure mode, can you say more about how these requests were made ? Were they separate requests, and were there more than one flow-ingester processes running who might have received them?

Still interested in this answer, but I've got test reproductions and fixes for all of the issues described above in my current working branch.

**Comment by psFried** [comment]
> To confirm the failure mode, can you say more about how these requests were made ? Were they separate requests, and were there more than one flow-ingester processes running who might have received them?

I sent a single request to the `/ingest` endpoint that contained multiple documents that each mapped to a different partition.  In other words, one transaction that's creating multiple partitions at the same time. Does that seem consistent with the issues you've found?

**Comment by jgraettinger** [comment]
Huh I'm unclear how this setup would have produced the original raced Etcd transaction.

Was there more than one flow-ingester running ? Is it possible that two requests were sent somehow ?
Or, could this have overlapped the period where it was mis-configured with the wrong broker etcd root ?

**Comment by psFried** [comment]
Yeah, unfortunately I didn't start good detailed notes until after the initial failure, so I could have gotten some bits wrong there. There were 3 replicas of flow-ingester running, although I used `kubectl port-forward` and it seemed to route alll the requests to the same pod. I don't think there were two concurrent requests, but there was a prior request that ingested a few events, and I can't remember if I killed curl during that or not, but I seem to remember doing that at some point.
